### ISSUE TITLE
[V2] Using SSR API on client side

### ISSUE THREAD

**Issue by PlanetIrata** [styled-components/styled-components#893]
I still can't find a way to extract styles _from client side_ with SC2. With SC1, I did it this way:

```
[render static component...]
let cssRules = StyleSheet.rules().map(rule => rule.cssText).join('\n');
```
When upgrading to SC2, I switch to the new SSR API:
```
let sheet = new ServerStyleSheet();
[render static component...]
let cssRules = sheet.getStyleTags();
```
But since I'm on client side, I get the "BrowserTag cannot be cloned!" error when trying to create the ServerStyleSheet object (see #811).

Digging the source code, I find this way to do it:

```
[render static component...]
let cssRules = StyleSheet.instance.toHTML();
```
But it generates several `<styles>` tags, some with `data-styled-components` attributes, some comments...not a very clean stylesheet.

What is the proper way to extract SC `<styles>` from the client side, with or without SSR API ?

**Comment by philpl** [comment]
i'm not sure why you're trying to extract styles on the client-side? Anyway, @geelen can probably help

**Comment by PlanetIrata** [comment]
I'm using React to build a static page editor/generator, so I need at some point to extract the HTML (with React renderStatic...) and the styles (with SC). At this stage I do everything on the client side. In the end, I may split the app with a server, or not: I may embed the app in an Electron like wrapper, so I'll still need to extract the style from the client side.

**Comment by langovoi** [comment]
I have similar problem with used of [phemonic](https://github.com/phenomic/phenomic) (1.0.0-alpha.3).

In entry point i have next:
```jsx
const routes = () => (
  sheet.collectStyles(
    <Router history={browserHistory}>
      <Route path='/' component={require('./components/Home/index').default} />
    </Router>
  )
)

export default createApp(routes, Html)
```

This code execute on client and on server side.

**Comment by geelen** [comment]
Hmm, you're right, and a side-effect of the perf work I'm doing this should suddenly start working. But your workaround sounds like it's also working?

The data attributes and comments are necessary for SC to rehydrate, same as the react checksum and stuff. But potentially we could add a `getStaticCSS()` method (similar to [`renderToStaticMarkup`](https://facebook.github.io/react/docs/react-dom-server.html#rendertostaticmarkup)) to handle the case where you're not sending the bundle down at all...

**Comment by PlanetIrata** [comment]
Yes, the `StyleSheet.instance.toHTML()` workaround works, for now, but for long term, I think that a `getStaticCSS()` api would be cleaner.

**Comment by philpl** [comment]
@geelen I'd suggest us to go with getGlobalCSS instead. The naming is clearer as to what it does, and it will avoid confusion with future changes in v3, as "static CSS" will likely be used to refer to extracted CSS

**Comment by geelen** [comment]
Hmm, I was mainly trying to parrot React's API already. However, I wonder if we can detect whether we're in `renderToString` or `renderToStaticMarkup` and so we don't need a second entry point, `getStyleTags()` always returns whatever's rendered, but we don't render the data attributes or comments in a static markup render.

I'll have a look to see if that's possible.

**Comment by langovoi** [comment]
Hey @geelen, do you have any news?

**Comment by Igorbek** [comment]
I'm also experiencing this issue when trying to test with jest.
- styled-components@2.10
- jest@20
  - jest environment set to `jsdom` (with `node` it works just fine)

Exception 'BrowserTag cannot be cloned!' thrown on `new ServerStyleSheet()`

**Comment by Igorbek** [comment]
Repro code:
```ts
import * as React from 'react';
import styled, { ServerStyleSheet } from 'styled-components';

it('jsdom fails', () => {
    styled.div` color: red; `;
    new ServerStyleSheet();
});
```

**Comment by bhovhannes** [comment]
For me `StyleSheet.instance.toHTML();` doesn't work.

Is there any other workaround for "BrowserTag cannot be cloned!" error when used client-side?

**Comment by Igorbek** [comment]
For my case, which is rendering on server for testing in jsdom environment, I found the following workaround:
```ts
import StyleSheet from 'styled-components/lib/models/StyleSheet';

StyleSheet.reset(true); // reset to use server stylesheet
```

**Comment by damassi** [comment]
@Igorbek - would you mind posting a complete example? I'm testing via Mocha but not sure exactly how to implement your fix.

**Comment by Igorbek** [comment]
Sure, it is pretty simple:

```js
import { create } from 'react-test-renderer';
import { ServerStyleSheet } from 'styled-components';
import StyleSheet from 'styled-components/lib/models/StyleSheet';
import * as css from 'css';

StyleSheet.reset(true); // reset to use server stylesheet

export function testComponent(element) {
    const sheet = new ServerStyleSheet();
    const output = create(sheet.collectStyles(element));

    expect(output).toMatchSnapshot(); // take snapshot of output

    // prettify css output slightly
    const cssStyles = sheet.getStyleElement()
        .map(el => el.props.dangerouslySetInnerHTML.__html)
        .join();
    const styles = css.stringify(css.parse(cssStyles));

    expect(styles).toMatchSnapshot(); // take a snapshot of produced css
}

// somewhere in the test:
import { testComponent } from ...

it('renders', () => {
    testComponent(<MyStyledComponentToTest ... />);
});
```

In fact, I have my own snapshot serializer that produce nice snapshots of input element, its rendered output and captured css.

**Comment by mCzolko** [comment]
@PlanetIrata please could you show us an example of `StyleSheet.instance.toHTML();` and how you use it. Me and @bhovhannes are not able to see the rendered CSS. Thank you

**Comment by PlanetIrata** [comment]
Sure, here it is : 
```
import ReactDOMServer from 'react-dom/server';
import styleSheet from 'styled-components/lib/models/StyleSheet';

styleSheet.reset(false);
ReactDOMServer.renderToStaticMarkup( <YourStyledComponentsApp />);
console.log(styleSheet.instance.toHTML());
```

**Comment by mxstbr** [comment]
Between @Igorbek's workaround and https://github.com/styled-components/jest-styled-components I think we've got this pretty well covered, I doubt we'll be spending any extra time on making our new SSR API work client-side.

Closing this!

**Comment by PlanetIrata** [comment]
Trying to upgrade from 2.4.0 to 3.1.5 and my app which use the "SSR client-side" workaround I posted here (2 comments above) don't build anymore : ```Module not found: Can't resolve 'styled-components/lib/models/StyleSheet'```

Haven't find anything new in the doc regarding SSR rendering with the V3 on client side (Electron, etc...), still the same SC2 sample with ServerStyleSheet throwing the same error when used client side (```Unhandled Promise Rejection: Error: BrowserTag cannot be cloned!```)

Reverting to SC 2.4.0 but any help would be appreciated, thanks.

**Comment by damassi** [comment]
@PlanetIrata - there's a breaking change in the 3.0 release: https://www.styled-components.com/releases#v3.0.0 / v3.0.1_shipping-only-flat-bundles -- that path is no longer valid.

**Comment by kitten** [comment]
@damassi We're continuing a discussion on the breaking change here: https://github.com/styled-components/styled-components/issues/1487

tl;dr the path is gone but it's still being exposed as a secret internal
